### PR TITLE
Support for backends with defective qubits and gates

### PR DESCRIPTION
**Pull request by 1ucian0**
Like #4083 but assuming the functional coupling map is connected. ~This approach is simpler, so the only pending here deceside a coupling map when is disconnected.~ (feature added in 232e1b8c8)

Example:
Consider the following coupling map, were `Q1` is non-operational:
```
         0 ↔ (1) ↔ 3 ↔ 4
              ↕
              2
```
The functional coupling map in this situation is only ` 3 ↔ 4`.
```python
circuit = QuantumCircuit(QuantumRegister(2, 'qr'))
circuit.h(range(2))
circuit.cz(0, 1)
circuit.measure_all()
result = transpile(circuit, backend=FakeOurenseFaultyQ1())
print(result)
```
```
disconnected_0 -> 0 ──────────────────────────────────────

      faulty_0 -> 1 ──────────────────────────────────────

disconnected_1 -> 2 ──────────────────────────────────────
                    ┌──────────┐                  ░ ┌─┐
          qr_0 -> 3 ┤ U2(0,pi) ├──■───────────────░─┤M├───
                    └──────────┘┌─┴─┐┌──────────┐ ░ └╥┘┌─┐
          qr_1 -> 4 ────────────┤ X ├┤ U2(0,pi) ├─░──╫─┤M├
                                └───┘└──────────┘ ░  ║ └╥┘
            meas_0: ═════════════════════════════════╩══╬═
                                                        ║
            meas_1: ════════════════════════════════════╩═
```

### CODE REVIEW

**Comment by 1ucian0** [comment]
This is an example choosing the largest connected component:
```
         0 ↔ 1 (↔) 3 ↔ 4
             ↕
             2
```
The largest connected functional coupling map is `0 ↔ 1 ↔ 2`.
```python
circuit = QuantumCircuit(QuantumRegister(3, 'qr'))
circuit.h(range(3))
circuit.ccx(0, 1, 2)
circuit.measure_all()
result = transpile(circuit,
                   backend=self.backend,
                   optimization_level=level,
                   seed_transpiler=42)
print(result)
```
```
                    ┌──────────┐                                                                    ┌───┐     ┌─────────────┐                   ░       ┌─┐
          qr_0 -> 0 ┤ U2(0,pi) ├───────────────────────────────────■────────────────────────────────┤ X ├──■──┤ U2(0,5pi/4) ├───────────────────░───────┤M├
                    ├──────────┤                       ┌───┐     ┌─┴─┐┌──────────┐┌───┐┌───────────┐└─┬─┘┌─┴─┐└─────────────┘ ┌──────────┐      ░ ┌─┐   └╥┘
          qr_1 -> 1 ┤ U2(0,pi) ├──■─────────────────■──┤ X ├──■──┤ X ├┤ U1(pi/4) ├┤ X ├┤ U1(-pi/4) ├──■──┤ X ├───────■────────┤ U1(pi/4) ├──■───░─┤M├────╫─
                    └──────────┘┌─┴─┐┌───────────┐┌─┴─┐└─┬─┘┌─┴─┐└───┘└──────────┘└─┬─┘└┬──────────┤     └───┘     ┌─┴─┐     ┌┴──────────┤┌─┴─┐ ░ └╥┘┌─┐ ║
          qr_2 -> 2 ────────────┤ X ├┤ U1(-pi/4) ├┤ X ├──■──┤ X ├───────────────────■───┤ U1(pi/4) ├───────────────┤ X ├─────┤ U1(-pi/4) ├┤ X ├─░──╫─┤M├─╫─
                                └───┘└───────────┘└───┘     └───┘                       └──────────┘               └───┘     └───────────┘└───┘ ░  ║ └╥┘ ║
disconnected_0 -> 3 ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╫──╫──╫─
                                                                                                                                                   ║  ║  ║
disconnected_1 -> 4 ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╫──╫──╫─
                                                                                                                                                   ║  ║  ║
            meas_0: ═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╩══╬══╬═
                                                                                                                                                      ║  ║
            meas_1: ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╩══╬═
                                                                                                                                                         ║
            meas_2: ═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╩═
```

**Comment by ajavadia** [comment]
In this example (https://github.com/Qiskit/qiskit-terra/pull/4110#issuecomment-611267864) what happens if the 1<-->2 link is defective?

Also what happens in general if a user specifies `initial_layout` in the transpiler? I think it should raise and not attempt to change the provided initial layout behind the scenes.

**Comment by 1ucian0** [comment]
> In this example ([#4110 (comment)](https://github.com/Qiskit/qiskit-terra/pull/4110#issuecomment-611267864)) what happens if the 1<-->2 link is defective?

Same circuit? Does not fit.
```python
class FakeOurenseFaulty(FakeOurense):
    """
         0 ↔ 1 (↔) 3 ↔ 4
            (↕)
             2
    """
    def properties(self):
        with open("./qiskit/test/mock/backends/ourense/props_ourense.json", "r") as f_prop:
            props = json.load(f_prop)
        for qubit in props['gates']:
            if qubit['name'] == 'cx1_3' or qubit['name'] == 'cx3_1' or qubit['name'] == 'cx2_1' or qubit['name'] == 'cx1_2':
                qubit['parameters'].append({"date": "2000-01-01 00:00:00Z",
                                            "name": "operational",
                                            "unit": "",
                                            "value": 0})
        return BackendProperties.from_dict(props)

circuit = QuantumCircuit(QuantumRegister(3, 'qr'))
circuit.h(range(3))
circuit.ccx(0, 1, 2)
circuit.measure_all()
result = transpile(circuit,
                   backend=FakeOurenseFaulty(),
                   optimization_level=1)
```
```
TranspilerError: 'Number of qubits (3) in circuit3 is greater than maximum (2) in the coupling_map'
```
 
> Also what happens in general if a user specifies `initial_layout` in the transpiler? I think it should raise and not attempt to change the provided initial layout behind the scenes.

Hmm... I think if `initial_layout` is in terms of the working coupling map, it should work (it does not currently because a bug).

For example when `Q1` is non-operational
```
         0 ↔ (1) ↔ 3 ↔ 4
              ↕
              2
```
The functional coupling map in this situation is only ` 3 ↔ 4`. If `initial_layout=[3,4]` should work, right?

```python
circuit = QuantumCircuit(QuantumRegister(2, 'qr'))
circuit.h(range(2))
circuit.cx(0, 1)
circuit.measure_all()
result = transpile(circuit, backend=FakeOurenseFaultyQ1(), optimization_level=1, initial_layout=[3,4])
```

**Comment by ajavadia** [comment]
I meant *only* 1–2 defective, what happens?
Yes if the initial layout is compatible with the functional Coupling map it should work. If not it should raise, without attempting to change the initial layout

**Comment by 1ucian0** [comment]
> I meant *only* 1–2 defective, what happens?

```
                    ┌──────────┐                                                                                                        ┌──────────┐      ░ ┌─┐
          qr_0 -> 0 ┤ U2(0,pi) ├───────────────────────────────────■──────────────────────────────────■────────────────────────■────────┤ U1(pi/4) ├──■───░─┤M├──────
                    ├──────────┤                       ┌───┐     ┌─┴─┐┌──────────┐┌───┐┌───────────┐┌─┴─┐     ┌───┐          ┌─┴─┐     ┌┴──────────┤┌─┴─┐ ░ └╥┘┌─┐
          qr_1 -> 1 ┤ U2(0,pi) ├──■─────────────────■──┤ X ├──■──┤ X ├┤ U1(pi/4) ├┤ X ├┤ U1(-pi/4) ├┤ X ├──■──┤ X ├──■───────┤ X ├─────┤ U1(-pi/4) ├┤ X ├─░──╫─┤M├───
                    └──────────┘  │                 │  └─┬─┘  │  └───┘└──────────┘└─┬─┘└───────────┘└───┘  │  └─┬─┘  │       └───┘     └───────────┘└───┘ ░  ║ └╥┘
disconnected_0 -> 2 ──────────────┼─────────────────┼────┼────┼─────────────────────┼──────────────────────┼────┼────┼───────────────────────────────────────╫──╫────
                                ┌─┴─┐┌───────────┐┌─┴─┐  │  ┌─┴─┐                   │   ┌──────────┐     ┌─┴─┐  │  ┌─┴─┐┌─────────────┐                   ░  ║  ║ ┌─┐
          qr_2 -> 3 ────────────┤ X ├┤ U1(-pi/4) ├┤ X ├──■──┤ X ├───────────────────■───┤ U1(pi/4) ├─────┤ X ├──■──┤ X ├┤ U2(0,5pi/4) ├───────────────────░──╫──╫─┤M├
                                └───┘└───────────┘└───┘     └───┘                       └──────────┘     └───┘     └───┘└─────────────┘                   ░  ║  ║ └╥┘
     ancilla_0 -> 4 ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────░──╫──╫──╫─
                                                                                                                                                          ░  ║  ║  ║
            meas_0: ═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╩══╬══╬═
                                                                                                                                                                ║  ║
            meas_1: ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╩══╬═
                                                                                                                                                                   ║
            meas_2: ═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╩═
```

**Comment by 1ucian0** [comment]
> Yes if the initial layout is compatible with the functional Coupling map it should work. If not it should raise, without attempting to change the initial layout

done in cee7c366b1ce6e0bfd60b91a0e44249f2152aaaa

File: qiskit/test/mock/backends/__init__.py [hunk 1]
```diff
@@ -21,6 +21,9 @@
 from .yorktown import FakeYorktown
 from .tenerife import FakeTenerife
 from .ourense import FakeOurense
+from .ourense import FakeOurenseFaultyQ1
+from .ourense import FakeOurenseFaultyCX13
+from .ourense import FakeOurenseFaultyCX01
```
Comment (ajavadia) [inline qiskit/test/mock/backends/__init__.py:26]: I would not add these to this directory. Just build them in the tests that you need. Trying to keep this directory a snapshot of real backends (i've been meaning to move them somewhere other than qiskit/test/)

File: qiskit/transpiler/coupling.py [hunk 2]
```diff
@@ -308,6 +308,10 @@ def from_grid(cls, num_rows, num_columns, bidirectional=True):
                     cmap.add_edge(node, right)
         return cmap
 
+    def largest_connected_component(self):
+        """Return a set of qubtis in the largest connected component """
```
Comment (ajavadia) [inline qiskit/transpiler/coupling.py:312]:
```suggestion
        """Return a set of qubits in the largest connected component."""
```

File: qiskit/providers/models/backendproperties.py [hunk 3]
```diff
@@ -314,6 +314,25 @@ def gate_property(self,
...
+            qubits: The qubit to find the operational status for.
+
+        Returns:
+            bool: Operational status of the given gate. True if the gate is operational,
+            False otherwise.
+        """
+        properties = self.gate_property(gate, qubits)
+        if 'operational' in properties:
+            return bool(properties['operational'][0])
+        return True  # if property operational not existant, then True
```
Comment (ajavadia) [inline qiskit/providers/models/backendproperties.py:334]:
```suggestion
        return True  # if property operational not existent, then True.
```

File: qiskit/providers/models/backendproperties.py [hunk 4]
```diff
@@ -416,6 +435,21 @@ def readout_error(self, qubit: int) -> float:
...
+        Args:
+            qubit: Qubit for which to return operational status of.
+
+        Returns:
+            Operational status of the given qubit.
+        """
+        properties = self.qubit_property(qubit)
+        if 'operational' in properties:
+            return bool(properties['operational'][0])
+        return True  # if property operational not existant, then True
```
Comment (ajavadia) [inline qiskit/providers/models/backendproperties.py:451]:
```suggestion
        return True  # if property operational not existent, then True.
```

File: qiskit/test/mock/backends/__init__.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/test/mock/backends/__init__.py]: Done in 13b92f3af

File: qiskit/providers/basebackend.py [hunk 5]
```diff
@@ -62,6 +62,30 @@ def configuration(self):
         """
         return self._configuration
 
+    def faulty_qubits(self):
```
Comment (ajavadia) [inline qiskit/providers/basebackend.py:65]: this should not be a method of base backend. it should be in `qiskit/providers/models/backendproperties.py`

File: qiskit/providers/basebackend.py [hunk 6]
```diff
@@ -62,6 +62,30 @@ def configuration(self):
...
+        properties = self.properties()  # pylint: disable=assignment-from-none
+        faulty = []
+        if properties is None:
+            return faulty
+        for qubit in range(self._configuration.n_qubits):
+            if not properties.is_qubit_operational(qubit):
+                faulty.append(qubit)
+        return faulty
+
+    def faulty_gates(self):
```
Comment (ajavadia) [inline qiskit/providers/basebackend.py:77]: same here

File: qiskit/providers/basebackend.py (hunk 5 above)
Comment (1ucian0) [reply qiskit/providers/basebackend.py:65]: Moved in db52ca406

File: qiskit/providers/basebackend.py (hunk 6 above)
Comment (1ucian0) [reply qiskit/providers/basebackend.py:77]: Moved in db52ca406

**Pull request merged**